In [1]:
#loading libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
#reading csv file into dataframe
df = pd.read_csv('ks-projects-201612.csv', encoding = "latin")

/Users/tatis/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (13,14,15) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [3]:
#accessing the file 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 323750 entries, 0 to 323749
Data columns (total 17 columns):
ID                323750 non-null int64
name              323746 non-null object
category          323745 non-null object
main_category     323750 non-null object
currency          323750 non-null object
deadline          323750 non-null object
goal              323750 non-null object
launched          323750 non-null object
pledged           323750 non-null object
state             323750 non-null object
backers           323750 non-null object
country           323750 non-null object
usd pledged       319960 non-null object
Unnamed: 13       625 non-null object
Unnamed: 14       12 non-null object
Unnamed: 15       4 non-null object
Unnamed: 16       1 non-null float64
dtypes: float64(1), int64(1), object(15)
memory usage: 42.0+ MB


#### deadline and lauch are not in the time format,  few numbers columns aren't integer/floats, 4 last columns are inconsistent and have very few information

In [4]:
df.head()

ID                                               name   \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000004038                                     Where is Hank?   
2  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
3  1000011046  Community Film Project: The Art of Neighborhoo...   
4  1000014025                               Monarch Espresso Bar   

        category  main_category  currency             deadline   goal   \
0          Poetry     Publishing       GBP  2015-10-09 11:36:00   1000   
1  Narrative Film   Film & Video       USD  2013-02-26 00:20:50  45000   
2           Music          Music       USD  2012-04-16 04:24:11   5000   
3    Film & Video   Film & Video       USD  2015-08-29 01:00:00  19500   
4     Restaurants           Food       USD  2016-04-01 13:38:27  50000   

             launched  pledged       state  backers  country  usd pledged   \
0  2015-08-11 12:12:28        0      failed        0       GB            0   
1  2013-01-12 00:20:50      220      failed        3       US          220   
2  2012-03-17 03:24:11        1      failed        1       US            1   
3  2015-07-04 08:35:03     1283    canceled       14       US         1283   
4  2016-02-26 13:38:27    52375  successful      224       US        52375   

  Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  
0         NaN         NaN         NaN          NaN  
1         NaN         NaN         NaN          NaN  
2         NaN         NaN         NaN          NaN  
3         NaN         NaN         NaN          NaN  
4         NaN         NaN         NaN          NaN

#### is the time of launch important? maybe makes sence to extract the date only as well as for dealine

In [7]:
#df['Unnamed: 13'].unique() #gives bunch of numbers, impossible to figure out waht are those; will drop 4 "Unnamed" columns

In [5]:
df.duplicated().sum() #no duplicates found, nice! 

0

In [8]:
df.sample(5)

ID                                               name   \
299357    85361299                                   Learn WordPress!   
227141   420118988                                       Sleeve Clips   
118947  1706199229  Help Lauren Get to Art Book Club's September S...   
195485   231155088                                 ArtMoose TV Series   
100006  1594151490  Daniel Boone and the Opening of the American West   

          category  main_category  currency             deadline   goal   \
299357     Software     Technology       USD  2013-04-14 18:46:41    900   
227141       Design         Design       USD  2014-03-16 02:37:42   3225   
118947   Public Art            Art       USD  2014-08-19 23:29:58    450   
195485   Television   Film & Video       USD  2014-06-16 07:30:00  12000   
100006  Documentary   Film & Video       USD  2014-02-16 21:05:59   7500   

                  launched   pledged       state  backers  country   \
299357  2013-03-15 17:46:41  19508.13  successful      678       US   
227141  2014-02-14 03:37:42  10647.62  successful      327       US   
118947  2014-08-07 23:29:58       615  successful       21       US   
195485  2014-05-01 21:06:51     12029  successful       70       US   
100006  2014-01-17 21:05:59        50      failed        1       US   

       usd pledged  Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  
299357     19508.13         NaN         NaN         NaN          NaN  
227141     10647.62         NaN         NaN         NaN          NaN  
118947          615         NaN         NaN         NaN          NaN  
195485        12029         NaN         NaN         NaN          NaN  
100006           50         NaN         NaN         NaN          NaN

In [9]:
df.columns #stupid space at the end of each column, need to rename 

Index(['ID ', 'name ', 'category ', 'main_category ', 'currency ', 'deadline ',
       'goal ', 'launched ', 'pledged ', 'state ', 'backers ', 'country ',
       'usd pledged ', 'Unnamed: 13', 'Unnamed: 14', 'Unnamed: 15',
       'Unnamed: 16'],
      dtype='object')

In [10]:
df_c1 = df.copy() #copying df 

In [11]:
#removing the space by replace it with nothing :D 
df_c1 = df_c1.rename(columns = lambda x: x.replace(' ', ''))

In [12]:
df_c1.columns #much better 

Index(['ID', 'name', 'category', 'main_category', 'currency', 'deadline',
       'goal', 'launched', 'pledged', 'state', 'backers', 'country',
       'usdpledged', 'Unnamed:13', 'Unnamed:14', 'Unnamed:15', 'Unnamed:16'],
      dtype='object')

In [13]:
df_c1.category.unique()

array(['Poetry', 'Narrative Film', 'Music', 'Film & Video', 'Restaurants',
       'Food', 'Drinks', 'Product Design', 'Documentary', 'Nonfiction',
       'Indie Rock', 'Crafts', 'Games', 'Design', 'Comic Books',
       'Art Books', 'Fashion', 'Theater', 'Comics', 'DIY', 'Webseries',
       'Food Trucks', 'Public Art', 'Illustration', 'Photography', 'Pop',
       'Tabletop Games', 'Art', 'Family', 'Fiction', 'Accessories',
       'Rock', 'Hardware', 'Weaving', 'Gadgets', 'Web', 'Jazz',
       'Ready-to-wear', 'Festivals', 'Video Games', 'Anthologies',
       'Publishing', 'Shorts', 'Electronic Music', 'Radio & Podcasts',
       'Apps', 'Cookbooks', 'Apparel', 'Metal', 'Hip-Hop', 'Periodicals',
       'Dance', 'Technology', 'Painting', 'Software', 'World Music',
       'Photobooks', 'Drama', 'Architecture', 'Young Adult',
       'Mobile Games', 'Flight', 'Fine Art', 'Action', 'Playing Cards',
       'Makerspaces', 'Punk', "Children's Books", 'People', 'Ceramics',
       'Fabrication Tool

In [14]:
df_c1.category.nunique() # ouch, so many

770

In [15]:
df_c1.main_category.nunique() #that's better

120

In [16]:
df_c1.country.unique() #mm, what are those numbers instead of countries?

array(['GB', 'US', 'CA', 'NO', 'AU', 'IT', 'DE', 'IE', 'ES', 'N,"0', 'MX',
       'SE', 'FR', 'NL', 'NZ', 'CH', 'AT', 'BE', 'DK', 'HK', 'LU', '14',
       '46', '10', '0', '42', 'SG', '35', '780', '122', '213', '2', '3',
       '126', '73', '19', '1', '142', 'failed', '140', '38', '24', '60',
       '37', '29', '167', '45', '23', '4', '85', '9', '248', '111', '18',
       '13', '28', '40', '54', '17', '20', '195', '25', '72', '22', '12',
       '158', '62', '307', '31', '107', '7', '110', '6', '69', '68', '84',
       '75', '119', '78', '150', '77', '83', '89', '16', '154', '92',
       '11', '116', '56', '136', '58', '27', '26', '582', '43', '5', '41',
       '30', '177', '36', '15', '49', '51', '9854', '168', '8', '57',
       '283', '61', '33', '74', '157', '405', '91', '174', '813', '66',
       '52', '114', '288', '109', '21', '32', '169', '50', '98', '483',
       '289', '88', '124', '159', '222', '65', '631', '59', '53', '117',
       '197', 'successful', '70', '55', '108', '63'

In [18]:
df_c1.loc[df_c1['country'] == "78"]

ID                             name                  category  \
64523  1383969232  Venus Wrote a Hymn for our Baby   Pira Wants to Record It   

      main_category currency deadline                 goal launched  \
64523         Faith    Music      USD  2014-07-10 15:08:00      888   

                   pledged state     backers country usdpledged Unnamed:13  \
64523  2014-06-26 09:03:06  2759  successful      78         US       2759   

      Unnamed:14 Unnamed:15  Unnamed:16  
64523        NaN        NaN         NaN

In [19]:
df_c1.loc[df_c1['country'] == "174"]

ID                          name                 category  \
150479  189401226  Bird Language with Jon Young   an instructional video   

       main_category      currency deadline                 goal launched  \
150479   Documentary  Film & Video      USD  2011-09-16 23:20:31     6000   

                    pledged state     backers country usdpledged Unnamed:13  \
150479  2011-08-02 23:20:31  7839  successful     174         US       7839   

       Unnamed:14 Unnamed:15  Unnamed:16  
150479        NaN        NaN         NaN

#### ok, it looks like where there was more than one category, the extra one was pushed in a wrong column and thus all the other entries were moved to the right; now it's clear what are those Unnamed columns with numbers. basically, we get to move it one (or more for some rows) column back 

In [20]:
df_c1.loc[df_c1['country'] == 'successful']

ID                              name      category  \
225692  410977509   Handmade Artisan Soaps: Natural        Gentle   
305503  890269949  Community Clothing. Make Clothes   Create Jobs   

              main_category currency deadline goal             launched  \
225692   Good for your skin   Crafts   Crafts  USD  2015-10-08 22:08:39   
305503        Restore Pride  Fashion  Fashion  GBP  2016-03-16 16:24:27   

       pledged                state  backers     country usdpledged  \
225692     200  2015-09-08 22:08:39      435  successful         10   
305503   75000  2016-02-15 17:24:27  88619.2  successful       1020   

       Unnamed:13 Unnamed:14 Unnamed:15  Unnamed:16  
225692         US        435        NaN         NaN  
305503         GB     128535        NaN         NaN

In [21]:
df_c1.loc[df_c1['country'] == '2015-09-10 18:15:45']

ID          name  category main_category    currency  \
269970  677103185  SixSixSeven    Angels        Demons    Religion    

         deadline            goal launched pledged                state  \
269970   Esoteric  Graphic Novels   Comics     USD  2015-10-10 01:00:00   

       backers              country usdpledged Unnamed:13 Unnamed:14  \
269970     750  2015-09-10 18:15:45         25     failed          1   

       Unnamed:15  Unnamed:16  
269970         US        25.0

In [22]:
df_c1.loc[df_c1['country'] == 'failed']

ID                      name  \
13795   1081139420        The Rolling Stones   
16482   1097082409                 Pixel art   
61447   1365419849                       VIA   
112934  1670090635  EXCALIBRE:Sword of Truth   
243349   516522480            American Booty   
275186   708058934                   Be Wise   

                                  category            main_category  \
13795                      BEGGARS BANQUET   50 Years in the Making   
16482                         personalized        Retro Gaming art.   
61447    ME (V.isual I.nteractive A.rchive       M.ercury E.dition)   
112934                       Path of Light       Kingdom of Heaven.   
243349        USA-made products Series One                  Spirits   
275186                         Wear Wisdom            Divine Wisdom   

          currency    deadline goal             launched pledged  \
13795         Rock       Music  USD  2011-08-12 01:17:48    4625   
16482       Crafts      Crafts  USD  2016-07-28 21:36:59     500   
61447     Software  Technology  USD  2016-03-19 20:40:23   10000   
112934  Nonfiction  Publishing  USD  2016-04-01 08:00:00   24000   
243349    Painting         Art  USD  2013-02-02 03:45:14    1776   
275186     Apparel     Fashion  USD  2015-02-27 13:53:56   10000   

                      state backers country usdpledged Unnamed:13 Unnamed:14  \
13795   2011-06-04 01:17:48      20  failed          2         US         20   
16482   2016-06-28 21:36:59       0  failed          0         US          0   
61447   2016-02-18 21:40:23      26  failed          2         US         26   
112934  2016-02-22 21:57:30       1  failed          1         US          1   
243349  2013-01-03 03:45:14     401  failed          7         US        401   
275186  2015-01-28 13:53:56       1  failed          1         US          1   

       Unnamed:15  Unnamed:16  
13795         NaN         NaN  
16482         NaN         NaN  
61447         NaN         NaN  
112934        NaN         NaN  
243349        NaN         NaN  
275186        NaN         NaN

#### so if we look at initial amount of non-null objects in all Unnamed columns, it's 625+12+4+1 which makes it 642 or 0.2% of total amount of entries, which means we can easily just drop those

In [23]:
#make a list of real countries, and if country isn't in the list, entry will be dropped
country_list = df_c1.country.unique().tolist()

In [24]:
df_c1.country.unique()

array(['GB', 'US', 'CA', 'NO', 'AU', 'IT', 'DE', 'IE', 'ES', 'N,"0', 'MX',
       'SE', 'FR', 'NL', 'NZ', 'CH', 'AT', 'BE', 'DK', 'HK', 'LU', '14',
       '46', '10', '0', '42', 'SG', '35', '780', '122', '213', '2', '3',
       '126', '73', '19', '1', '142', 'failed', '140', '38', '24', '60',
       '37', '29', '167', '45', '23', '4', '85', '9', '248', '111', '18',
       '13', '28', '40', '54', '17', '20', '195', '25', '72', '22', '12',
       '158', '62', '307', '31', '107', '7', '110', '6', '69', '68', '84',
       '75', '119', '78', '150', '77', '83', '89', '16', '154', '92',
       '11', '116', '56', '136', '58', '27', '26', '582', '43', '5', '41',
       '30', '177', '36', '15', '49', '51', '9854', '168', '8', '57',
       '283', '61', '33', '74', '157', '405', '91', '174', '813', '66',
       '52', '114', '288', '109', '21', '32', '169', '50', '98', '483',
       '289', '88', '124', '159', '222', '65', '631', '59', '53', '117',
       '197', 'successful', '70', '55', '108', '63'

In [26]:
df_c1.loc[df_c1['country'] == 'N,"0'] #what's this country? looks like nothing wrong with the
                                     #rest of the info, so we leave it in the list of valid countries for now

ID                                               name  \
150     1000694855                      STREETFIGHTERZ WHEELIE MURICA   
287      100149523                        Duncan Woods - Chameleon EP   
549     1003023003          The Making of Ashley Kelley's Debut Album   
561     1003130892                       Butter Side Down Debut Album   
650     1003629045                            Chase Goehring debut EP   
714     1004013077           Spencer Capier Instrumental Project 2015   
730     1004126342  LUKAS LIGETI'S 50TH BIRTHDAY FESTIVAL: ORIGINA...   
747     1004255433      The Battle For Breukelen: A Neighborhood Epic   
751     1004298993  "Tamboura Plays Violin" - a collection of Pop ...   
765     1004402863        Nightingale Noel - An A Cappella Holiday CD   
883     1005185256                             Local Music Connection   
891      100522240  DO NOT DUPLICATE - Selected by 2015 Devon Film...   
963     1005653464                                       SA-4 Studios   
1073    1006327667                           Letters to the Wild - EP   
1447    1008671527                          The Fitness Mindset   :-)   
1448    1008675685               Inspired & the Sleep EP - Eyelid Kid   
1475    1008815806  Into Winter: An Instrumental Christmas Experiment   
1531    1009154868       Help advance scientific guitar & bass design   
1561    1009312309                                        Mr. Pickles   
1624    1009774693                Universal Peace | Universal Justice   
1685    1010217550                               Solarboot-Havel-Tour   
1769    1010694326                                          Surrender   
1893    1011508551                               EBENEZER Solo Record   
1978    1011952280                         NHSOB Summer Festival 2015   
1998    1012058337                                        Degobah Max   
2051    1012414277    WE ARE WHO WE ARE BUT WHATEVER WE WANT TO BE!!!   
2097    1012704793                               Life In The Bus Lane   
2102    1012744036          An Oratorio for our Time - Last Stop Cafe   
2122    1012807897                                  Juliet Remembered   
2232    1013389538  Make admission to the London Music Awards cere...   
...            ...                                                ...   
320925   982862687                                    Bydlo - A Novel   
320937   982919439                                              Faith   
321068   983728691                   The Octopodes' 5th Studio Album!   
321288   985231591                                           TWO FISH   
321473   986253191  DEVELOP MAGAZINE FOR THE ARTS, MUSIC, AND DESI...   
321513   986471034     Help Revival finish their new album, Guidance.   
321655    98744618                         Vinyl for Dancing Pigeons!   
321896   988763792    Maud Sings Maud: A Musical Betsy-Tacy Companion   
321973   989218116                   Keep on Movin' - The Debut Album   
322005   989499289                                          Sojourner   
322065   989874415                        Love and Soul through Sound   
322321   991409513                          Chris Copeland's First EP   
322333   991483230  Support Vocal Synergy's Second Semester and Te...   
322382   991820324                                  The Blend CD 2015   
322472   992433603                                  Jonny Come Lately   
322730   994080962                                    Maximum Coppage   
322818   994621918                                        Star Sailor   
322881   994911312                    "Songwriter assistance" service   
322883   994933706                             Whiskey and Holy Water   
323077    99613274               Double EP and Music Video Fundraiser   
323198   996908566                  The World Needs More Big Thinkers   
323275    99739483                                       No More Fear   
323313   997554968  Susan Picking: Making Music with Kids in the S...   
3

In [27]:
country_list = ['GB', 'US', 'CA', 'NO', 'AU', 'IT', 'DE', 'IE', 'ES', 'N,"0', 'MX','SE', 'FR', 'NL', 'NZ', 'CH', 'AT', 'BE', 'DK', 'HK', 'LU']

In [45]:
df_c2 = df_c1[df_c1['country'].isin(country_list)]

In [46]:
df_c2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 322999 entries, 0 to 323749
Data columns (total 17 columns):
ID               322999 non-null int64
name             322995 non-null object
category         322999 non-null object
main_category    322999 non-null object
currency         322999 non-null object
deadline         322999 non-null object
goal             322999 non-null object
launched         322999 non-null object
pledged          322999 non-null object
state            322999 non-null object
backers          322999 non-null object
country          322999 non-null object
usdpledged       319209 non-null object
Unnamed:13       0 non-null object
Unnamed:14       0 non-null object
Unnamed:15       0 non-null object
Unnamed:16       0 non-null float64
dtypes: float64(1), int64(1), object(15)
memory usage: 44.4+ MB


In [47]:
# now we can drop all those Unnamed columns cause they are empty
df_c2 = df_c2.drop(columns = ['Unnamed:13', 'Unnamed:14', 'Unnamed:15', 'Unnamed:16'], axis = 1)

In [48]:
df_c2.sample()

ID         name category main_category currency  \
46617  1277107509  Gaming Vlog   Shorts  Film & Video      AUD   

                  deadline   goal             launched pledged   state  \
46617  2014-12-18 23:59:17  10000  2014-11-18 23:59:17       0  failed   

      backers country usdpledged  
46617       0      AU          0

In [49]:
#let's take care of time data
df_c2.deadline[0]

'2015-10-09 11:36:00'

In [50]:
df_c2['deadline'] = df_c2['deadline'].str.split(' ').str.get(0)
df_c2['launched'] = df_c2['launched'].str.split(' ').str.get(0)

In [51]:
df_c2.sample()

ID                                           name category  \
309968  917345297  Whare no name - the new home of M?ori Theatre  Theater   

       main_category currency    deadline   goal    launched pledged  \
309968       Theater      NZD  2015-04-21  10000  2015-03-23   10355   

             state backers country     usdpledged  
309968  successful      61      NZ  7834.34023445

In [52]:
df_c2['deadline'] = pd.to_datetime(df_c2['deadline'])
df_c2['launched'] = pd.to_datetime(df_c2['launched'])

In [53]:
df_c2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 322999 entries, 0 to 323749
Data columns (total 13 columns):
ID               322999 non-null int64
name             322995 non-null object
category         322999 non-null object
main_category    322999 non-null object
currency         322999 non-null object
deadline         322999 non-null datetime64[ns]
goal             322999 non-null object
launched         322999 non-null datetime64[ns]
pledged          322999 non-null object
state            322999 non-null object
backers          322999 non-null object
country          322999 non-null object
usdpledged       319209 non-null object
dtypes: datetime64[ns](2), int64(1), object(10)
memory usage: 44.5+ MB


In [54]:
df_c2.sample()

ID                     name      category main_category  \
256763  597783808  Skydiving with Rob Ford  Film & Video  Film & Video   

       currency   deadline goal   launched pledged   state backers country  \
256763      USD 2015-06-04  250 2015-05-05       3  failed       2      US   

       usdpledged  
256763          3

In [ ]:
#and turn columns with numbers to integers 
df_c2[["goal", "pledged", "backers", "usdpledged"]] = df_c2[["goal", "pledged", "backers", "usdpledged"]].apply(pd.to_numeric)

In [ ]:
df_c2.sample()

#### next step is to plot few graphs to see if there any correlations

In [ ]:
import plotly.plotly as py
import plotly.graph_objs as go